docs/decisions/0002-feature-based-application-organization.rst

```diff
@@ -0,0 +1,157 @@
+
+**2. Create strict module boundaries**
+
+A feature should have a public interface exposed via an index.js file in the feature directory.  Consumers of a feature should limit themselves to importing only from the public exports.
+
+Modules are configured from above. For react components, this is giving them props, for services, this is giving it apiClient/configuration, etc.
+
+**3. Avoid circular dependencies**
+
+Circular dependencies are unresolvable in webpack, and will result in something being imported as 'undefined'.  They're also incredibly difficult and frustrating to track down.  Properly factoring your features and supporting modules should help avoid these sorts of issues.  In general, a feature should never need to import from its parent, or from its lineage (grandparents, great-grandparents, etc.).  Furthermore, think about how "general purpose" a given feature or module is.  If you find yourself importing from a domain-specific feature in your general utility module, then something is probably ill-factored.
+
+Secondary principles
+====================
+
+**A. Separate business logic from components**
+
+In order to isolate our view layer (react) from business logic inherent in our global state, APIs, and side effect management, we want to adopt the "ducks" organization (see references).  This involves isolating business logic into a
+sub directory of a feature.  See "B" below for an example.
+
+**B. Files in a feature directory are named by type of code**
+
+Since directory names describe features, file names describe types of code.
+
+::
+
+  /profile
+    /index.js // public interface
+    /ProfilePage.jsx // view
+    /ProfilePhotoUploader.jsx // supporting view
+    /ducks // business logic
+      /actions.js
+      /reducers.js
+      /selectors.js
+
+Within the ducks directory, if multiple files of a type are necessary, first consider whether there is really a sub-feature here that should be defined.  If no discernable sub-feature is emerging, then turn the file (e.g., reducers.js) into a directory (reducers/)
+and put multiple files in it, named by what the reducer does (reducers/AccountReducer.js).  The directory can then have an index.js file which documents the reducers that can be imported.  Consider again whether there's really a sub-feature here.
+
+Multiple components and reducers:
+
+::
+
+  /profile
+    /index.js // public interface
+    /ProfilePage.jsx
+    /ProfilePhotoUploader.jsx
+    /Avatar.jsx
+    /ducks
+      /actions.js
+      /reducers
+        /index.js
+        /FormReducers.js
+        /ProfilePhotoReducers.js
```

Comment:
Is it necessary to provide a model example of having multiple instances of Reducers files within a feature?  If someone does this, should they perceive this as a code-smell of extracting out a separate sub-feature?

Reply:
It's a vey good point - I wasn't 100% sure if I thought it was a code-smell, but I think it's likely you're right.